### ISSUE TITLE
CDI injection not available on metrics bean in micro PAYARA-4174

### ISSUE THREAD

**Issue by tiesebarrell** [payara/Payara#3881]
# Description #
----------

When defining microprofile metrics on an ApplicationScoped bean, CDI injection into that bean does not work on Payara Micro. An exception is thrown internally when the metrics 
are gathered by the /metrics endpoint for Gauges. Any metrics that use an injected bean to report their values are excluded from the results due to this exception. An example exception caught by debugging shows that the injected proxy to the CDI bean no longer refers to a valid target.

## Expected Outcome

The metrics bean should be able to use a CDI bean to provide the value of a Gauge metric.

## Current Outcome

org.jboss.weld.exceptions.IllegalStateException: WELD-000229: Contextual reference of Producer Method [NumberGenerator] with qualifiers [@Default @Any @Named] declared as [[BackedAnnotatedMethod] @Produces @Named @ApplicationScoped public test.NumberGeneratorProducer.numberGenerator()] is not valid after container test.war.bda shutdown

The example above uses a Producer for the bean, but the same behaviour is observed for regular beans defined as ApplicationScoped within the same application, even though the exception message is slightly different.

This only happens on ApplicationScoped beans as far as known. For JAX-RS resources for instance, as used in many examples showing the metrics API, the injection works correctly and can be used with direct requests to the resouce as well as when the metrics are requested from /metrics.

The full distribution has no issues with this setup. From the spec, it is also not clear this is to be an expected limitation, which the support in the full edition corroborates. It occurs on the micro edition in versions 5.184 and 5.191.

## Steps to reproduce (Only for bug reports) 

Deploy application with Gauge metrics. Call /metrics to aggregate the metrics.

### Samples

```
@ApplicationScoped
public class MetricsResource {

    @Inject
    @Native
    private NumberGenerator numberGenerator;

    @Inject
    private NumberGenerator namedNumberGenerator;

    @Gauge(name = "numberGenerated", unit = MetricUnits.NONE)
    public int numberGenerated() {
        return numberGenerator.generate();
    }

    @Gauge(name = "namedNumberGenerated", unit = MetricUnits.NONE)
    public int namedNumberGenerated() {
        return namedNumberGenerator.generate();
    }

    @Gauge(name = "numberGeneratedInline", unit = MetricUnits.NONE)
    public int numberGeneratedInline() {
        return new Random().nextInt();
    }

    public void init(@Observes @Initialized(ApplicationScoped.class) Object init) {
        System.out.println("Initialising MetricsResource");
    }
}
```
Calling /metrics on Payara Micro results in the internal exception mentioned and erases both Gauges that use an injected bean from the results. The inline Gauge is still in the result. All three are included when running the same on Payara Full.

## Environment ##

- **Payara Version**: 5.184 and 5.191
- **Edition**: Micro
- **JDK Version**: 8.x OpenJDK
- **Operating System**: Windows (any)
- **Database**: n/a

**Comment by rdebusscher** [comment]
Hi @tiesebarrell, 

This issues doesn't seems to be related to the HealthCheck one but more a timing problem.

So can you try by disabling the CDI deployment in parallel by specifying the following command in the _postbootcommand_ file.

`configs.config.server-config.cdi-service.enable-concurrent-deployment=false`

see this docmentation page about the postboot command file https://docs.payara.fish/documentation/payara-micro/asadmin/pre-and-post-boot-scripts.html.

**Comment by tiesebarrell** [comment]
@rdebusscher Thanks, we will investigate if this solves the issue in our case.

**Comment by tiesebarrell** [comment]
It seems that configuration either doesn't exist or it's not working: 

reference-application-service_1  | [2019-05-02T07:44:37.938+0000] [] [WARNING] [] [fish.payara.boot.runtime.BootCommand] [tid: _ThreadID=1 _ThreadName=main] [timeMillis: 1556783077938] [levelValue: 900] Boot Command configs.config
.server-config.cdi-service.enable-concurrent-deployment=false failed Exception while executing command.

Where can I find an overview of the commands that can be used?

**Comment by smillidge** [comment]
probably should be
`set configs.config.server-config.cdi-service.enable-concurrent-deployment=false`

**Comment by tiesebarrell** [comment]
Thanks @smillidge , that made sure the configuration was correctly applied. However, the behaviour observed is still the same. Injection on the bean that produces the metric fails, which results in the metric being left out of the results.

**Comment by OndroMih** [comment]
This issue was also reported as a customer ticket. We can reproduce the problem and will work on a fix.

**Comment by OndroMih** [comment]
This is now fixed in version 5.194 (and also in the version 5.191.9 available to customers with a commercial support)
